### PR TITLE
Add groove/ridge borders to new border path.

### PR DESCRIPTION
**Pull request by glennw**
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/1147)

### CODE REVIEW

File: webrender/res/ps_border_corner.vs.glsl [hunk 1]
```diff
@@ -3,26 +3,36 @@
...
-    }
+    return invalid;
+}
+
+void set_radii(float style,
+               vec2 radii,
+               vec2 widths,
+               vec2 adjusted_widths) {
+    vRadii0.xy = get_radii(radii, 2.0 * widths);
+    vRadii0.zw = get_radii(radii - widths, vec2(-widths));
```
Comment (kvark) [inline webrender/res/ps_border_corner.vs.glsl]: nit: vec2() is unnecessary

File: webrender/res/ps_border_corner.vs.glsl [hunk 2]
```diff
@@ -32,6 +42,34 @@ void set_edge_line(vec2 border_width,
...
 
+void write_color(vec4 color0, vec4 color1, float style, vec2 delta) {
+    vec4 modulate = vec4(1.0 + 0.3 * delta.x,
+                         1.0 - 0.3 * delta.x,
+                         1.0 + 0.3 * delta.y,
+                         1.0 - 0.3 * delta.y);
+
+    switch (int(style)) {
+        case BORDER_STYLE_GROOVE:
+            vColor00 = vec4(color0.rgb * modulate.y, color0.a);
```
Comment (kvark) [inline webrender/res/ps_border_corner.vs.glsl]: nit: could be rewritten by turning the match to produce a `vec4 modulate` vector, which is then used to compute `vColorXY`

File: webrender/res/ps_border_edge.vs.glsl [hunk 3]
```diff
@@ -7,52 +7,97 @@ void write_edge_distance(float p0,
...
+            vAlphaSelect = 0.0;
+            break;
+        default:
+            vAlphaSelect = 1.0;
+            break;
+    }
+}
+
+void write_color(vec4 color, float style, bool flip) {
+    vec2 modulate;
```
Comment (kvark) [inline webrender/res/ps_border_edge.vs.glsl]: nit: `vec2 modulate =  flip ? ... : ... ;`

File: webrender/res/ps_border_edge.vs.glsl [hunk 4]
```diff
@@ -7,52 +7,97 @@ void write_edge_distance(float p0,
...
+    if (flip) {
+        modulate = vec2(0.7, 1.3);
+    } else {
+        modulate = vec2(1.3, 0.7);
+    }
+
+    switch (int(style)) {
+        case BORDER_STYLE_GROOVE:
+            vColor0 = vec4(color.rgb * modulate.y, color.a);
```
Comment (kvark) [inline webrender/res/ps_border_edge.vs.glsl]: similarly, could be simplified by leaving only the `modulate` computation in the match

File: webrender/res/ps_border_corner.fs.glsl [hunk 5]
```diff
@@ -97,15 +99,22 @@ void main(void) {
...
+        // SDF subtract main radii
+        float d_main = max(d0, 0.5 * afwidth - d1);
 
-        // SDF union
-        float d = min(d_outer, d_inner);
+        // SDF subtract inner radii (double style borders)
+        float d_inner = max(d2 - 0.5 * afwidth, -d3);
+
+        // Select how to combine the SDF based on border style.
+        float d = mix(max(d_main, -d_inner), d_main, vSDFSelect);
```
Comment (kvark) [inline webrender/res/ps_border_corner.fs.glsl]: I've long lost track of the math side of the issue...

**Comment by glennw** [comment]
@kvark Thanks - fixed the CI issue and comments up. Once I complete the other border styles, I'll write up a document describing how each of the shaders works :)

**Pull request merged**